File: python/ics/cobraOps/defineBenchGeometry.py

import numpy as np

from loadCfgXml import loadCfgXml


def defineBenchGeometry(centers, useRealMaps, useRealLinks):
    """Defines the bench geometry.
    
    thetaDIR is > 0 for positive move out of home (opposite of phi)    
    thetaDIR is < 0 for negative move out of home (same as phi)
    
    output is a structure with
        center (Mx1)
        L1     (Mx1)
        L2     (Mx1)
        phiIn  (Mx1)
        phiOut (Mx1)
        tht0   (Mx1)
        NNmap  (MxM), logical
        rf
        distCobras
        minDist
        pids 
        mids 
        S1Nm 
        S1Pm, S2Pm, S2Nm, binWidth);
        
    """
    KeepOutAngle = 0.1  # radians (~ 5.5 deg)
    alpha = 0.07
    beta = 0.50
    pids = np.array([], dtype="int")
    mids = np.array([], dtype="int")
    mapAssignment = None

    # This is an epsilon to make sure that values near zero in theta are
    # intepreted on the positive (or negative) side of the cut,
    # respectively. Make it negative for same same direction moveouts (positive
    # for positive hardstop).
    #
    # tht1 implemented below is an opposite-sense hard stop.  when
    # that's working, theta direction will be specified on a per cobra
    # basis, not on a full instrument basis, as is implemented here.
    thteps = 2e-10  # vesitigial

    if useRealMaps or useRealLinks:
        # Load the configuration file
        cobraConfig = loadCfgXml("updatedMotorMapsFromThisRun2.xml")
        
        # Get the arm data containers
        dataContainers = cobraConfig.findall("ARM_DATA_CONTAINER")
        nDataContainers = len(dataContainers)

        # Get the number of steps from the first data container        
        slowCalibrationTable = dataContainers[0].find("SLOW_CALIBRATION_TABLE")
        
        if slowCalibrationTable.find("Joint1_fwd_stepsizes") is not None:
            nSteps = int(slowCalibrationTable.find("Joint1_fwd_stepsizes").text.split(",")[0])
        else:
            nSteps = 0

        # Save some of the configuration information
        pids = np.zeros(nDataContainers)
        mids = np.zeros(nDataContainers)
        cx = np.zeros(nDataContainers)
        cy = np.zeros(nDataContainers)
        tht0 = np.zeros(nDataContainers)  # hard stop angle for same sense move-out
        tht1 = np.zeros(nDataContainers)  # hard stop angle for opposite sense move-out
        phiIn = np.zeros(nDataContainers)
        phiOut = np.zeros(nDataContainers)
        L1 = np.zeros(nDataContainers)
        L2 = np.zeros(nDataContainers)
        S1Pm = np.zeros((nDataContainers, nSteps))
        S2Pm = np.zeros((nDataContainers, nSteps))
        S1Nm = np.zeros((nDataContainers, nSteps))
        S2Nm = np.zeros((nDataContainers, nSteps))

        for i in xrange(nDataContainers):
            dataContainer = dataContainers[i]
            
            header = dataContainer.find("DATA_HEADER")
            pids[i] = int(header.find("Positioner_Id").text)
            mids[i] = int(header.find("Module_Id").text)
            
            kinematics = dataContainer.find("KINEMATICS")
            cx[i] = float(kinematics.find("Global_base_pos_x").text)
            cy[i] = float(kinematics.find("Global_base_pos_y").text)
            tht0[i] = float(kinematics.find("CCW_Global_base_ori_z").text) * np.pi / 180
            tht1[i] = float(kinematics.find("CW_Global_base_ori_z").text) * np.pi / 180
            phiIn[i] = float(kinematics.find("Joint2_CCW_limit_angle").text) * np.pi / 180 - np.pi
            phiOut[i] = float(kinematics.find("Joint2_CW_limit_angle").text) * np.pi / 180 - np.pi
                        
            if kinematics.find("Link1_Link_Length") is not None:
                L1[i] = float(kinematics.find("Link1_Link_Length").text)
            else:
                L1[i] = 25.0

            if kinematics.find("Link2_Link_Length") is not None:
                L2[i] = float(kinematics.find("Link2_Link_Length").text)
            else:
                L2[i] = 25.0
            
            slowCalibrationTable = dataContainer.find("SLOW_CALIBRATION_TABLE")

            if slowCalibrationTable.find("Joint1_fwd_stepsizes") is not None:
                # Regions are expected to be standard bin width of 3.6deg
                # phi out of HS in positive direction
                # tht out of HS in negative direction
                Joint1Fwd = slowCalibrationTable.find("Joint1_fwd_stepsizes").text.split(",")[2:-1]
                Joint2Fwd = slowCalibrationTable.find("Joint2_fwd_stepsizes").text.split(",")[2:-1]
                Joint1Rev = slowCalibrationTable.find("Joint1_rev_stepsizes").text.split(",")[2:-1]
                Joint2Rev = slowCalibrationTable.find("Joint2_rev_stepsizes").text.split(",")[2:-1]
                S1Pm[i] = [3.6 / float(val) for val in Joint1Fwd]
                S2Pm[i] = [3.6 / float(val) for val in Joint2Fwd]
                S1Nm[i] = [3.6 / float(val) for val in Joint1Rev]
                S2Nm[i] = [3.6 / float(val) for val in Joint2Rev]

        # Calculated map ranges (not physical range of motion).  other
        # option is to read it directly from regions.
        map_range = {}
        map_range["tht"] = np.array([0, nSteps * 3.6 * np.pi / 180])
        map_range["phi"] = np.array([0, nSteps * 3.6 * np.pi / 180]) - np.pi

        phiOut -= KeepOutAngle
        phiIn = np.maximum(phiIn, -np.pi)  # HACK SOLUTION TO BAD PHI HOMES [2016-07-15]
        phiIn += KeepOutAngle
        center = cx + cy * 1j

        # Physical parameters
        rf = 1000.0 / 90  # fiber holder radius
        distCobras = 8.0 * 1000.0 / 90  # distance between cobras for custom geometries.
        minDist = 2 * rf  # minimum distance between fiber and anything
        
        # variables defined in this branch that can propagate to different centers:
        # L1
        # L2
        # phiIn
        # phiOut
        # S[12][FR]m
        configData = {}
        configData["pids"] = pids
        configData["mids"] = mids
        configData["phiIn"] = phiIn
        configData["phiOut"] = phiOut
        configData["L1"] = L1
        configData["L2"] = L2
        configData["S1Pm"] = S1Pm
        configData["S2Pm"] = S2Pm
        configData["S1Nm"] = S1Nm
        configData["S2Nm"] = S2Nm
        configData["distCobras"] = distCobras

    if centers is not None:
        # Cobra center locations
        center = centers
        csize = len(center)
        ONE = np.ones(csize)
    
        # Physical parameters
        rf = 1.0  # fiber holder radius
        distCobras = 8.0  # distance between cobras for custom geometries.
        minDist = 2 * rf  # minimum distance between fiber and anything

        thtrange = 385 * np.pi / 180  # range of motion of the theta stage.

        tht0 = 2 * np.pi * np.random.random(csize)  # same sense hard stop
        tht1 = (tht0 + thtrange) % (2 * np.pi)  # opp sense hard stop

        if useRealMaps:
            nDataContainers = configData["S1Pm"].shape[0]
            nSteps = configData["S1Pm"].shape[1]
            S1Pm = np.zeros((csize, nSteps))
            S2Pm = np.zeros((csize, nSteps))
            S1Nm = np.zeros((csize, nSteps))
            S2Nm = np.zeros((csize, nSteps))
            
            mapAssignment = (nDataContainers * np.random.random((4, csize))).astype("int")
            
            S1Pm = configData["S1Pm"][mapAssignment[0]]
            S2Pm = configData["S2Pm"][mapAssignment[1]]
            S1Nm = configData["S1Nm"][mapAssignment[2]]
            S2Nm = configData["S2Nm"][mapAssignment[3]]
        else:
            S1Pm = np.array([])
            S2Pm = np.array([])
            S1Nm = np.array([])
            S2Nm = np.array([])
        
        if useRealLinks:
            # take L1,L2, phiIn/Out from CobraConfig.
            pix2mm = distCobras / configData["distCobras"]
            nDataContainers = configData["L1"].shape[0]

            indexes = (nDataContainers * np.random.random((4, csize))).astype("int")
            
            phiIn = configData["phiIn"][indexes[0]] 
            phiOut = configData["phiOut"][indexes[1]] 
            L1 = configData["L1"][indexes[2]] * pix2mm              
            # ## There was a bug here it was using L1
            L2 = configData["L2"][indexes[3]] * pix2mm              
        else: 
            DeltaPhi = np.pi - 2 * KeepOutAngle  # throw of the phi arm
            phiOut = -KeepOutAngle * (1.0 + 0.2 * np.random.random(csize))
            phiIn = phiOut - DeltaPhi

            linkLength = 2.375
            L1 = linkLength * np.ones(csize)
            L2 = linkLength * np.ones(csize)

    rMin = np.abs(L1 + L2 * np.exp(1j * phiIn))
    rMax = np.abs(L1 + L2 * np.exp(1j * phiOut))
    
    binWidth = 2 * np.pi / 100;  # this should be defined higher up in the code.[PHM 20160901
    
    # parameters for populating the annular patrol areas uniformly
    dA = 1.0 / ((rMax / rMin) ** 2 - 1.0)  # fractional keepout area
    rRange = np.sqrt(rMax ** 2 - rMin ** 2)

    # by default, phi home is phi in.  existence of mat files *may* change phi home.
    phiHome = phiIn.copy()

    # for the realities of the test bench, having phi too far in is
    # inconvenient.  Use a larger value so that the theta arm does not
    # go crazy.
    phiHome += 0.5

    # read in phiHome from elsewhere
    """
    # TODO!!!
    matfiles = ls2cell('*.mat')
    
    if len(matfiles) > 0:
        print('Warning -- using mat file positions for home in defineBenchGeometry')
        data = loadTestData('', CobraConfig, 1) 
        
        for id = 1:numel(data):
            if ~isempty(data(id).xst):
                kk = find(pids == id)
                phiHome(kk) = mean(data(id).s2.startAngle(1,:)) * pi / 180 - pi
    """
    
    # home positions (0 = ss, 1 = os)
    home0 = center + L1 * np.exp(1j * tht0) + L2 * np.exp(1j * (tht0 + phiHome))
    home1 = center + L1 * np.exp(1j * tht1) + L2 * np.exp(1j * (tht1 + phiHome))

    # theta overlap
    tht_overlap = ((tht1 - tht0 + np.pi) % (2 * np.pi)) - np.pi;

    # generate the nearest neighbor map/structure
    csize = len(center)
    epsilon = 1e-9  # small number to take care of roundoff errors.
    CCdist = np.zeros((csize, csize))
    
    for i in xrange(csize):
        CCdist[i] = np.abs(center - center[i])
        
    distCent = 2 * np.median(L1 + L2) * (8.0 / 9.5)
    nnMap = np.logical_and(CCdist < 1.5 * distCent , CCdist > epsilon)
    (row, col) = np.where(nnMap)
    # inUpperTriangle = row < col
    NN = {}
    NN["row"] = row  # (inUpperTriangle)
    NN["col"] = col  # (inUpperTriangle)
    NN["xy"] = (center[row] + center[col]) / 2

    # Calculate some parameters for the field geometry
    field = {}
    field["cm"] = np.mean(center)
    field["R"] = np.max(np.abs(center - field["cm"]) + rMax)
    # Fit a line to the centers and calculate the slope
    x = np.real(center)
    y = np.imag(center)
    slope = (csize * (x * y).sum() - x.sum() * y.sum()) / (csize * (x * x).sum() - x.sum() ** 2);
    field["ang"] = np.arctan([slope])

    output = {}
    output["center"] = center
    output["L1"] = L1
    output["L2"] = L2
    output["phiIn"] = phiIn
    output["phiOut"] = phiOut
    output["tht0"] = tht0
    output["tht1"] = tht1
    output["rMin"] = rMin
    output["rMax"] = rMax
    output["dA"] = dA
    output["rRange"] = rRange
    output["home0"] = home0
    output["home1"] = home1
    output["nnMap"] = nnMap
    output["NN"] = NN
    output["rf"] = rf
    output["distCobras"] = distCobras
    output["minDist"] = minDist
    output["S1Nm"] = S1Nm
    output["S1Pm"] = S1Pm
    output["S2Pm"] = S2Pm
    output["S2Nm"] = S2Nm
    output["map_range"] = map_range
    output["binWidth"] = binWidth
    output["alpha"] = alpha
    output["beta"] = beta
    output["pids"] = pids
    output["mids"] = mids
    output["thteps"] = thteps
    output["tht_overlap"] = tht_overlap
    output["field"] = field
    output["map"] = mapAssignment

    return output

File: matlab/collisions/defineBenchGeometry.m
function output=defineBenchGeometry(centers, useRealMaps, useRealLinks)
% define bench geometry 
% call defineBenchGeometry([],1,1) to read the config file directly.
%
% thetaDIR is > 0 for positive move out of home (opposite of phi)
% thetaDIR is < 0 for negative move out of home (same as phi)
%
% output is a structure with
%    center (Mx1)
%    L1     (Mx1)
%    L2     (Mx1)
%    phiIn  (Mx1)
%    phiOut (Mx1)
%    tht0   (Mx1)
%    NNmap  (MxM), logical
%    rf
%    distCobras
%    minDist
%    pids 
%    mids 
%    S1Nm 
%    S1Pm, S2Pm, S2Nm, binWidth);

KeepOutAngle = 0.1; % radians (~ 5.5 deg)

alpha = 0.07;
beta = 0.50;
pids = [];
mids = [];

% this is an epsilon to make sure that values near zero in theta are
% intepreted on the positive (or negative) side of the cut,
% respectively. Make it negative for same same direction moveouts (positive
% for positive hardstop).
%
% tht1 implemented below is an opposite-sense hard stop.  when
% that's working, theta direction will be specified on a per cobra
% basis, not on a full instrument basis, as is implemented here.
thteps = 2e-10;  %% vesitigial


if useRealMaps | useRealLinks
    CobraConfig = loadCfgXml;
    cx = [];
    cy = [];
    tht0 = []; % hard stop angle for same sense move-out
    tht1 = []; % hard stop angle for opposite sense move-out
    phiIn = [];
    phiOut = [];
    L1 = [];
    L2 = [];
    S1Nm = []; % slow maps
    S2Pm = [];
    S1Pm = [];
    S2Nm = [];
    F1Nm = []; % fast maps
    F2Pm = [];
    F1Pm = [];
    F2Nm = [];
    for ii = 1:length(CobraConfig.ARM_DATA.ARM_DATA_CONTAINER)
        dc = CobraConfig.ARM_DATA.ARM_DATA_CONTAINER{ii};
        pids = [pids; str2num(dc.DATA_HEADER.Positioner_Id.Text)];
        mids = [mids; str2num(dc.DATA_HEADER.Module_Id.Text)];
        cx = [cx; str2num(dc.KINEMATICS.Global_base_pos_x.Text)];
        cy = [cy; str2num(dc.KINEMATICS.Global_base_pos_y.Text)];
        tht0 = [tht0; (str2num(dc.KINEMATICS.CCW_Global_base_ori_z.Text))*pi/180];
        tht1 = [tht1; (str2num(dc.KINEMATICS.CW_Global_base_ori_z.Text))*pi/180];
        phiIn  = [phiIn ; (str2num(dc.KINEMATICS.Joint2_CCW_limit_angle.Text))*pi/180 - pi];
        phiOut = [phiOut; (str2num(dc.KINEMATICS.Joint2_CW_limit_angle.Text)) *pi/180 - pi];
        if isfield(dc.KINEMATICS, 'Link1_Link_Length')
            L1 = [L1; str2num(dc.KINEMATICS.Link1_Link_Length.Text)];
        else
            L1 = [L1; 25];
        end
        if isfield(dc.KINEMATICS, 'Link2_Link_Length')
            L2 = [L2; str2num(dc.KINEMATICS.Link2_Link_Length.Text)];
        else
            L2 = [L2; 25];
        end
        if isfield(dc.SLOW_CALIBRATION_TABLE, 'Joint1_fwd_stepsizes')
            % regions are expected to be standard bin width of 3.6deg
            % phi out of HS in positive direction
            % tht out of HS in negative direction
            values = str2num(dc.SLOW_CALIBRATION_TABLE.Joint1_fwd_stepsizes.Text);
            values = 1./values * 3.6;
            S1Pm = [S1Pm; values(3:end)];
            values = str2num(dc.SLOW_CALIBRATION_TABLE.Joint2_fwd_stepsizes.Text);
            values = 1./values * 3.6;
            S2Pm = [S2Pm; values(3:end)];
            values = str2num(dc.SLOW_CALIBRATION_TABLE.Joint1_rev_stepsizes.Text);
            values = 1./values * 3.6;
            S1Nm = [S1Nm; values(3:end)];
            values = str2num(dc.SLOW_CALIBRATION_TABLE.Joint2_rev_stepsizes.Text);
            values = 1./values * 3.6;
            S2Nm = [S2Nm; values(3:end)];
        end
        if isfield(dc.FAST_CALIBRATION_TABLE, 'Joint1_fwd_stepsizes')
            % regions are expected to be standard bin width of 3.6deg
            % phi out of HS in positive direction
            % tht out of HS in negative direction
            values = str2num(dc.FAST_CALIBRATION_TABLE.Joint1_fwd_stepsizes.Text);
            values = 1./values * 3.6;
            F1Pm = [F1Pm; values(3:end)];
            values = str2num(dc.FAST_CALIBRATION_TABLE.Joint2_fwd_stepsizes.Text);
            values = 1./values * 3.6;
            F2Pm = [F2Pm; values(3:end)];
            values = str2num(dc.FAST_CALIBRATION_TABLE.Joint1_rev_stepsizes.Text);
            values = 1./values * 3.6;
            F1Nm = [F1Nm; values(3:end)];
            values = str2num(dc.FAST_CALIBRATION_TABLE.Joint2_rev_stepsizes.Text);
            values = 1./values * 3.6;
            F2Nm = [F2Nm; values(3:end)];
        end
    end
    % calculated map ranges (not physical range of motion).  other
    % option is to read it directly from regions.
    map_range.tht = [0 size(S1Pm,2) * 3.6 * pi / 180];
    map_range.phi = [0 size(S2Pm,2) * 3.6 * pi / 180] - pi;
    
    phiOut = phiOut - KeepOutAngle;
    phiIn  = max(phiIn, -pi); % HACK SOLUTION TO BAD PHI HOMES [2016-07-15]
    phiIn  = phiIn + KeepOutAngle;
    center = cx + 1i*cy;
    %% physical parameters
    rf = 1000/90; % fiber holder radius
    distCobras = 8.000 * 1000/90; % distance between cobras for custom geometries.
    minDist = 2 * rf; % minimum distance between fiber and anything
    %% variables defined in this branch that can propagate to
    %% different centers:
    % L1
    % L2
    % phiIn
    % phiOut
    % S[12][FR]m
    configData = packstruct(L1,L2,phiIn,phiOut,S1Nm,S1Pm,S2Pm,S2Nm,distCobras, pids, mids);
else
    map_range.tht = [0 , 112 * 3.6 * pi / 180];
    map_range.phi = [0 , 112 * 3.6 * pi / 180] - pi;
end

if ~isempty(centers)
    %% Cobra center locations
    center = complex(reshape(centers,[],1));
    csize = size(center);
    ONE = ones(csize);
    
    %% physical parameters
    rf = 1.000; % fiber holder radius
    distCobras = 8.000; % distance between cobras for custom geometries.
    minDist = 2 * rf; % minimum distance between fiber and anything

    thtrange = 385 * pi/180; % range of motion of the theta stage.

    tht0 = 2*pi*rand(csize); % same sense hard stop
    tht1 = mod(tht0 + thtrange, 2*pi); % opp sense hard stop
    
    if useRealMaps
        mapAssignment = ceil(rand(length(center),4) * length(configData.L1));
        S1Pm   = configData.S1Pm(mapAssignment(:,1),:);
        S1Nm   = configData.S1Nm(mapAssignment(:,2),:);
        S2Pm   = configData.S2Pm(mapAssignment(:,3),:);
        S2Nm   = configData.S2Nm(mapAssignment(:,4),:);
    else
        S1Pm = ONE*ones(1,112)*50;
        S2Pm = ONE*ones(1,112)*50;
        S1Nm = ONE*ones(1,112)*50;
        S2Nm = ONE*ones(1,112)*50;
    end
    if useRealLinks
        %take L1,L2, phiIn/Out from CobraConfig.
        pix2mm = distCobras/configData.distCobras;
        LasVegas = ceil(rand(length(center),4) * length(configData.L1));
        L1     = configData.L1(LasVegas(:,1)) * pix2mm;
        L2     = configData.L1(LasVegas(:,2)) * pix2mm;
        phiIn  = configData.phiIn(LasVegas(:,3));
        phiOut = configData.phiOut(LasVegas(:,4));
    else 
        linkLength = 2.375;
        L1 = linkLength * ONE;
        L2 = linkLength * ONE;
        
        DeltaPhi = pi - 2*KeepOutAngle; % throw of the phi arm
        phiOut = -KeepOutAngle * (1 + 0.2 * randn(csize));
        phiIn  = phiOut - DeltaPhi;
    end
end

rMin = abs(L1 + L2 .* exp(1i*phiIn));
rMax = abs(L1 + L2 .* exp(1i*phiOut));
binWidth =  2 * pi/100;  %$ this should be defined higher up in the
                         %code.[PHM 20160901

%% parameters for populating the annular patrol areas uniformly
dA     = 1./( (rMax./rMin).^2 - 1 );  % fractional keepout area
rRange = sqrt(rMax.^2 - rMin.^2);

%% by default, phi home is phi in.  existence of mat files *may*
%% change phi home.
phiHome = phiIn;

%% for the realities of the test bench, having phi too far in is
%% inconvenient.  Use a larger value so that the theta arm does not
%% go crazy.

phiHome = phiHome + 0.5;

%% read in phiHome from elsewhere
matfiles = ls2cell('*.mat');
if length(matfiles) > 0
    disp('Warning -- using mat file positions for home in defineBenchGeometry');
    data = loadTestData('',CobraConfig,1); 
    for id = 1:numel(data)
        if ~isempty(data(id).xst)
            kk = find(pids == id);
            phiHome(kk) = mean(data(id).s2.startAngle(1,:)) * pi / 180 - pi;
        end
    end
end

%% home positions (0 = ss, 1 = os)
home0 = center + L1 .* exp(i*tht0) + L2 .* exp(i*(tht0 + phiHome));
home1 = center + L1 .* exp(i*tht1) + L2 .* exp(i*(tht1 + phiHome));

%% theta overlap
tht_overlap = mod(tht1 - tht0 + pi, 2*pi) - pi;

%% generate the nearest neighbor map/structure
epsilon = 1e-9;  % small number to take care of roundoff errors.
CCdist = abs(bsxfun(@minus, center, center.'));
distCent = 2 * median(L1 + L2) * (8.0 / 9.5);
nnMap = sparse(CCdist < 1.5*distCent & CCdist > epsilon);
[row col] = find(nnMap);
%inUpperTriangle = row < col;
NN.row = row;%(inUpperTriangle);
NN.col = col;%(inUpperTriangle);
NN.xy  = mean([center(NN.row), center(NN.col)],2);

%% calculate some parameters for the field geometry
field.cm = mean(center);
field.R  = max(abs(center - field.cm) + rMax);
field.ang = atan(subarray(cmplx(@polyfit,center,1),1));

output = packstruct(center, L1, L2, phiIn, phiOut, tht0, tht1, rMin, rMax, ...
                    dA, rRange,home0,home1,...
                    nnMap, NN, rf, distCobras, minDist, ...
                    S1Nm, S1Pm, S2Pm, S2Nm, map_range, binWidth, ...
                    alpha, beta, ...
                    pids, mids, thteps, tht_overlap, field);
if exist('mapAssignment','var')
    output.map = mapAssignment;
end

end
